### DISCUSSION TITLE
Improve building / dev env setup instructions

### DISCUSSION THREAD

**Discussion by pastk** [organicmaps/organicmaps#1538] [Q&A]
I'm willing to improve development environment setup / building documentation (Android-specific mostly), i.e. `docs/INSTALL.md`.
I have a few ideas and willing to propose PRs, however I need to clarify some points and get feedback first.

### Dedup the "getting sources"
At the moment "getting sources" sub-section is present in each target (Android, Desktop, iOS) in almost identical form.
I want to move it in a separate section preceding target-specific instructions.

### Android

1. What does the `./configure.sh` do and what is the difference between _open-source_ and _private_ builds?

2. Is it worthwhile to document the `splitApk` build param - and why its not used by default? And the `pch` (pre-compiled headers) param?

3. Document other build params like `-Parm64`

4. Add more details about studio setup, e.g. memory setting (I couldn't get the project open - it was stuck/frozen at "updating indexes" - until I increased studio's heap size from default 1.9Gb to 2.5Gb ), setting up virtual device or using physical one...

5. Could we lower `org.gradle.jvmargs=-Xmx8g` in `gradle.properties` to some minimally required setting? Newbies might have impression that 8Gb is the minimum to be able to build, however I found it builds fine using 3Gb (didn't try less yet). Those devs who have more RAM can increase this setting locally.

6. Does `org.gradle.parallel=true` actually needed and makes a difference? As far as I understand it makes a difference to multi-projects builds only (and OM is a single project build or am I mistaken here?) - at least that setting didn't do any difference for me.

7. Why suggest to always run `clean` before building (e.g. in `gradlew clean assembleXXX`)? My understanding is that it'll make a build slower by forcing to re-compile everything from scratch instead of re-building only the parts that were changed

8. Update build flavors, e.g. remove yandex, xiaomi, add fdroid...

9. Update `HOW_TO_USE_GRADLE.txt` and link to it from the `INSTALL.md`.

10. Maybe I'll add a separate sub-section with advice for people with low spec machines, e.g. how to conserve disk space and ram.

**Comment by biodranik** [comment]
1. `./configure.sh` sets up some important stuff for desktop and mobile builds. It could be a good idea to split what is required for mobile builds and desktop builds to make it easier. You can check this script and see what is going on inside. Private builds use private keys and configs used to publish apps in Google Play. Open-source builds do not use private keys and configs.
2. Please document `splitApk`. Why do you think it should be used by default? Regarding `pch` (Precompiled Headers), we didn't check it yet. Do you see any difference in compilation speed with and without it (after cleaning caches)?
3. Yes!
4. Yes!
5. No, but please mention it in the documentation.
6. Probably it's not needed. Do you see any speed difference with and without it (after cleaning all caches, of course)?
7. `clean` is not needed before every build, it looks like a historic artifact.
8. We'll update/reuse flavors later. It makes sense to publish the app everywhere.
9. Yes!
10. Good idea!

**Reply by pastk** [reply]
Thanks a lot for your feedback!

2. separate arch-specific apks take less space and are faster to download and process, but maybe its not that important for development; what is the current use case for `splitApk`? is it used to build releases only?
re `pch` - I'll test later and update
6. At the first glance it didn't change anything, but I didn't do any profiling or exact timing. ll test later more thoroughly and update.

**Reply by biodranik** [reply]
2. We don't use it now, apk split is just for convenience in some cases. Google Play split apk automatically. Uploading split releases can confuse users who don't really know the arch (some 64-bit devices may use 32-bit arch OS for example.

**Reply by pastk** [reply]
I did some tests for **parallel** and **pch** options.
The outcome is that **parallel** does nothing (so I suggest to remove it from `gradle.properties`) and **pch** gives 15% speed up (so I suggest to document it at least and maybe later after more testing make it default).

Setup:
_org.gradle.parallel=true_ and _org.gradle.jvmargs=-Xmx8g_ were commented out in `gradle.properties` (i.e. JVM memory setting was default)
I ran `./gradlew clean` before each run.

**Parallel-enabled build:**
`./gradlew -Parm64 --parallel --no-build-cache --rerun-tasks --profile assembleFdroidDebug`
run1: 20m 5s; run2: 20m 17s

**Parallel-disabled build:**
`./gradlew -Parm64 -Ppch --no-parallel --no-build-cache --rerun-tasks --profile assembleFdroidDebug`
run1: 20m 9s; run2: 20m 11s

**PCH-enabled build:**
`./gradlew -Parm64 -Ppch --no-parallel --no-build-cache --rerun-tasks --profile assembleFdroidDebug`
run1: 17m 2s; run2: 16m 59s
i.e. faster than non-pch by 3mins or 15%

All builds were successful but I didn't test-run the resulting apks.

**Reply by biodranik** [reply]
Cool, thanks! Let's remove parallel and document pch.

**Reply by pastk** [reply]
I also ran a few tests to see how JVM memory settings affect build times.
My conlusion is that a bigger JVM heap size doesn't affect the build speed - and it makes sense as 90+% of the build time is taken by cxx compilation which doesn't have anything to do with JVM. So the heap size should be big enough to run all the tasks, but there is no point to keep it higher.
So my suggestion is to reduce the current `-Xmx8g` setting to a `-Xmx1024m`.

**Details:**
Build cmd `./gradlew -Parm64 -Ppch --no-parallel --no-build-cache --rerun-tasks --profile assembleFdroidDebug`

By default (if there is no explicit setting in `gradle.properties`) gradle daemon is started with just `-Xms256m -Xmx512m`:
run1: 17m 2s; run2: 16m 59s

With a `org.gradle.jvmargs=-Xmx4g` in `gradle.properties`:
run1: 17m 1s; run2: 17m 19s

However with the default `-Xms256m -Xmx512m` setting the build failed for `FdroidRelease` at `Task :minifyFdroidReleaseWithR8`: `ERROR:R8: java.lang.OutOfMemoryError: Java heap space`.

So I increased it to 1Gb (`org.gradle.jvmargs=-Xmx1024m` in `gradle.properties`) and build was successful:
run1: 22m 18s

2Gb setting:
run1: 22m 11s

4Gb setting:
run1: 21m 53s

**Reply by biodranik** [reply]
Good, 1Gb should be enough.

**Reply by pastk** [reply]
For 1. `configure.sh` private mode I understand that the private repo is expected to contain only secure/private config files in question and the script just copies them over to the local "main" repository. Is that right?

**Reply by biodranik** [reply]
Correct.

**Comment by pastk** [comment]
A PR to Disable parallel builds and reduce JVM heap: https://github.com/organicmaps/organicmaps/pull/1548

**Reply by pastk** [reply]
A PR to Dedup and improve Getting Sources section: https://github.com/organicmaps/organicmaps/pull/1551

**Reply by pastk** [reply]
Add details and simplify the main flow: https://github.com/organicmaps/organicmaps/pull/1568

**Comment by pastk** [comment]
`android/HOW_TO_USE_GRADLE.txt` contains following:

```
@NOTE1: Run <organicmaps>/tools/android/update_assets.sh once every time when assets are added/removed
@NOTE2: To increase apk version, please edit gradle.properties
```
Are those notes relevant still? I.e. its not clear how to increase the apk version.

Other that those lines it seems that `android/HOW_TO_USE_GRADLE.txt` duplicates `docs/INSTALL.md`.
So my suggestion is to add missing bits into `docs/INSTALL.md` and remove `android/HOW_TO_USE_GRADLE.txt`.

**Reply by biodranik** [reply]
Version is automatically generated now, so the doc reference can be removed.

Script creates symlinks, so it should be actual now. Let's leave its doc here and also duplicate it in INSTALL.md or even move to CONTRIBUTE.md (as it's relevant to development, not for building/installing)

**Reply by pastk** [reply]
Hmm, but there is no `<organicmaps>/tools/android/update_assets.sh`?

**Reply by biodranik** [reply]
Right! It was removed… There is ./android/script/replace_links.bat left to workaround missing symlinks for git, but that is not the case anymore with working symlinks.

**Comment by pastk** [comment]
`cmake --build ../omim-build-debug --target help` is not useful or friendly to get a list of build targets.
First it fails if there were no build done previously. I fixed it by replacing with `tools/unix/build_omim.sh -d help`.

But anyway the output is >1000 lines of build targets still so not really readable or helpful...
Is it possible to reduce the output to "main" targets only? (I'm new to cmake..)
Or maybe we can compile a short list of really useful targets like _desktop_, _generator_tool_, etc.?

**Reply by biodranik** [reply]
Almost every folder and subfolder is a target. There should be a way to list cmake's high level projects/targets. Documenting is even better.
